src/libraries/Common/src/System/Runtime/InteropServices/ArrayMarshaller.cs

```diff
@@ -0,0 +1,227 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+//
+// Types in this file are used for generated p/invokes (docs/design/features/source-generator-pinvokes.md).
+// See the DllImportGenerator experiment in https://github.com/dotnet/runtimelab.
+//
+#if DLLIMPORTGENERATOR_INTERNALUNSAFE
+using Internal.Runtime.CompilerServices;
+#else
+using System.Runtime.CompilerServices;
+#endif
+using System.Diagnostics;
+
+namespace System.Runtime.InteropServices.GeneratedMarshalling
+{
+    internal unsafe ref struct ArrayMarshaller<T>
+    {
+        private T[]? managedArray;
```

Comment:
Use `_` prefix for fields? (libraries coding convention)

Reply:
Our plan is to go through the inserted code prior to merging into `main`. There are a lot of style issues we need to address.